### PR TITLE
Cherry-pick VACUUM stats & Vacuum progress related commits.

### PR DESCRIPTION
**Pull request by reshke**
Cherry-pick some GPDB commits.

This is part 1 of my native Yezzey support for CBDB plan https://github.com/yezzey-gp/yezzey/issues/40#issuecomment-2321149779

### CODE REVIEW

**Comment by reshke** [comment]
UPD: cherry-pick a6fdc27246e46bf3b72d43420c1454a37e68e273 from gp7 to fix incorrect regression test results

**Comment by lss602726449** [comment]
Thanks for your Merge Request. I think We'd better revert the last commit which have been merged and squash the last two commit together.

**Comment by reshke** [comment]
> Thanks for your Merge Request. I think We'd better revert the last commit which have been merged and squash the last two commit ([refac](https://github.com/cloudberrydb/cloudberrydb/pull/611/commits/5c3f686ae8fd544f8f6d14f60850d2300e3fa264) and [993d11e](https://github.com/cloudberrydb/cloudberrydb/commit/993d11ecc0a81b09977fc91e06302b0f5851923a)) together.

Hi, many thanks for your review.
did exacly as you suggested.

**Review by lss602726449 — APPROVED** [review]

File: src/backend/catalog/pg_attribute_encoding.c [hunk 1]
```diff
@@ -225,13 +225,16 @@ RelationGetAttributeOptions(Relation rel)
 	Datum *dats;
 	StdRdOptions **opts;
 	int i;
+	int	nvp;
+	
+	nvp = RelationGetNumberOfAttributes(rel);
```
[the remark this answers is not in the document]
Comment (reshke) [reply src/backend/catalog/pg_attribute_encoding.c]: thanks, done

**Comment by reshke** [comment]
Applied all review comments as suggested

**Review by my-ship-it — APPROVED** [review]

**Comment by ostinru** [comment]
hm... it seems that `make` breaks after this patch...

```
 > [cloudberry 8/8] RUN locale-gen en_US.utf8 \
 && ./configure --with-perl --with-python --with-libxml --with-gssapi --prefix=/usr/local/gpdb_src > /dev/null \
 && make -j8 > /dev/null \
 && make -j8 install > /dev/null

0.379 Generating locales (this might take a while)...
0.381   en_US.UTF-8... done
1.244 Generation complete.
29.97 appendonlyam.c: In function 'appendonly_insert':
29.98 appendonlyam.c:3052:9: error: unused variable 'firstSequence' [-Werror=unused-variable]
29.98  3052 |   int64 firstSequence = GetFastSequences(aoInsertDesc->segrelid,
29.98       |         ^~~~~~~~~~~~~
30.38 cc1: all warnings being treated as errors
30.38 make[4]: *** [<builtin>: appendonlyam.o] Error 1
30.38 make[4]: *** Waiting for unfinished jobs....
30.64 make[3]: *** [../../../src/backend/common.mk:39: appendonly-recursive] Error 2
30.64 make[3]: *** Waiting for unfinished jobs....
31.84 make[2]: *** [common.mk:39: access-recursive] Error 2
31.84 make[2]: *** Waiting for unfinished jobs....
34.40 make[1]: *** [Makefile:44: all-backend-recurse] Error 2
34.40 make: *** [GNUmakefile:11: all-src-recurse] Error 2
```


Good news - it works fine with
```
./configure --with-perl --with-python --with-libxml --with-gssapi --prefix=/usr/local/gpdb_src --enable-debug
```

**Comment by reshke** [comment]
> hm... it seems that `make` is broken (for me) after this patch...
> 
> ```
>  > [cloudberry 8/8] RUN locale-gen en_US.utf8 \
>  && ./configure --with-perl --with-python --with-libxml --with-gssapi --prefix=/usr/local/gpdb_src > /dev/null \
>  && make -j8 > /dev/null \
>  && make -j8 install > /dev/null
> 
> 0.379 Generating locales (this might take a while)...
> 0.381   en_US.UTF-8... done
> 1.244 Generation complete.
> 29.97 appendonlyam.c: In function 'appendonly_insert':
> 29.98 appendonlyam.c:3052:9: error: unused variable 'firstSequence' [-Werror=unused-variable]
> 29.98  3052 |   int64 firstSequence = GetFastSequences(aoInsertDesc->segrelid,
> 29.98       |         ^~~~~~~~~~~~~
> 30.38 cc1: all warnings being treated as errors
> 30.38 make[4]: *** [<builtin>: appendonlyam.o] Error 1
> 30.38 make[4]: *** Waiting for unfinished jobs....
> 30.64 make[3]: *** [../../../src/backend/common.mk:39: appendonly-recursive] Error 2
> 30.64 make[3]: *** Waiting for unfinished jobs....
> 31.84 make[2]: *** [common.mk:39: access-recursive] Error 2
> 31.84 make[2]: *** Waiting for unfinished jobs....
> 34.40 make[1]: *** [Makefile:44: all-backend-recurse] Error 2
> 34.40 make: *** [GNUmakefile:11: all-src-recurse] Error 2
> ```

Thanks for your feedback! will take a look soon

**Comment by reshke** [comment]
> hm... it seems that `make` is broken (for me) after this patch...
> 
> ```
>  > [cloudberry 8/8] RUN locale-gen en_US.utf8 \
>  && ./configure --with-perl --with-python --with-libxml --with-gssapi --prefix=/usr/local/gpdb_src > /dev/null \
>  && make -j8 > /dev/null \
>  && make -j8 install > /dev/null
> 
> 0.379 Generating locales (this might take a while)...
> 0.381   en_US.UTF-8... done
> 1.244 Generation complete.
> 29.97 appendonlyam.c: In function 'appendonly_insert':
> 29.98 appendonlyam.c:3052:9: error: unused variable 'firstSequence' [-Werror=unused-variable]
> 29.98  3052 |   int64 firstSequence = GetFastSequences(aoInsertDesc->segrelid,
> 29.98       |         ^~~~~~~~~~~~~
> 30.38 cc1: all warnings being treated as errors
> 30.38 make[4]: *** [<builtin>: appendonlyam.o] Error 1
> 30.38 make[4]: *** Waiting for unfinished jobs....
> 30.64 make[3]: *** [../../../src/backend/common.mk:39: appendonly-recursive] Error 2
> 30.64 make[3]: *** Waiting for unfinished jobs....
> 31.84 make[2]: *** [common.mk:39: access-recursive] Error 2
> 31.84 make[2]: *** Waiting for unfinished jobs....
> 34.40 make[1]: *** [Makefile:44: all-backend-recurse] Error 2
> 34.40 make: *** [GNUmakefile:11: all-src-recurse] Error 2
> ```

Does this PR make a fix for you https://github.com/cloudberrydb/cloudberrydb/pull/674 ?

**Comment by ostinru** [comment]
thanks! Build & test from `main` branch works fine now.

**Pull request merged**